python_modules/libraries/dagster-tableau/dagster_tableau/resources.py

```diff
@@ -0,0 +1,149 @@
+
+    def _fetch_json(
+        self,
+        endpoint: str,
+        data: Optional[Dict[str, Any]] = None,
+        method: str = "GET",
+        with_auth_header: bool = True,
+    ) -> Dict[str, Any]:
+        """Fetch JSON data from the Tableau API. Raises an exception if the request fails.
+
+        Args:
+            endpoint (str): The API endpoint to fetch data from.
+            data (Optional[Dict[str, Any]]): JSON-formatted data string to be included in the request.
+            method (str): The HTTP method to use for the request.
+            with_auth_header (bool): Whether to add X-Tableau-Auth header to the request. Enabled by default.
+
+        Returns:
+            Dict[str, Any]: The JSON data returned from the API.
+        """
+        response = self._make_request(
+            endpoint=endpoint, data=data, method=method, with_auth_header=with_auth_header
+        )
+        return response.json()
+
+    def _make_request(
+        self,
+        endpoint: str,
+        data: Optional[Dict[str, Any]] = None,
+        method: str = "GET",
+        with_auth_header: bool = True,
+    ) -> requests.Response:
+        headers = {
+            "accept": "application/json",
+            "content-type": "application/json",
+        }
+        if with_auth_header:
+            headers["X-tableau-auth"] = self._api_token
+        request_args: Dict[str, Any] = dict(
+            method=method,
+            url=f"{self.api_base_url}/{endpoint}",
+            headers=headers,
+        )
+        if data:
+            request_args["json"] = data
+        response = requests.request(**request_args)
+        response.raise_for_status()
+        return response
+
+    @cached_method
+    def get_workbooks(self) -> Dict[str, Any]:
+        """Fetches a list of all Tableau workbooks in the workspace."""
+        return self._fetch_json(self._with_site_id("workbooks"))
+
+    @cached_method
+    def get_workbook_data_sources(
+        self,
+        workbook_id: str,
+    ) -> Dict[str, Any]:
+        """Fetches a list of all data sources for a given workbook."""
+        return self._fetch_json(self._with_site_id(f"workbooks/{workbook_id}/connections"))
+
+    def sign_in(self):
+        """Sign in to the site in Tableau."""
+        data = {
+            "credentials": {
+                "personalAccessTokenName": self.personal_access_token_name,
+                "personalAccessTokenSecret": self.personal_access_token_value,
+                "site": {"contentUrl": self.site_name},
+            }
+        }
+        return self._fetch_json(
+            endpoint="auth/signin", data=data, method="POST", with_auth_header=False
+        )
+
+    def sign_out(self):
+        """Sign out from the site in Tableau."""
+        return self._make_request(endpoint="auth/signout", method="POST")
+
+    def _with_site_id(self, endpoint: str):
+        return f"sites/{self._site_id}/{endpoint}"
+
+
+class TableauCloudWorkspace(BaseTableauWorkspace):
```

Comment:
In our use case, the only difference between Tableau Cloud and Tableau Server is the API base URL. 

My first thought was to implement only one `TableauWorkspace` resource, with a parameter `host_name`, where users would pass the server name for Tableau Server and the domain for Tableau Cloud.

Reasons why I ended up implementing two resources:
- Some [Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api.htm) methods are exclusive to Tableau Cloud or Tableau Server.
- `server_name` and `pod_name` are part of Tableau's ontology, but `host_name` is not.

Reply:
@yuhan @benpankow - any thoughts here based on how we've implemented other resources?

Reply:
> Some Tableau API methods are exclusive to Tableau Cloud or Tableau Server

This seems like good motivation to keep the two separate in case we end up using these APIs. The bulk of the implementation here is shared code so I don't think it's too big a burden to keep both maintained.